Add Spin to Particle Container

cemitch99 opened the pull request:
The PR add the basic spin data structure and initialization.  It is needed by PR#1219.

- [x] add spin data structure(s) to particle container
- [x] add initialization of spin data
- [x] add user-facing inputs for spin tracking
- [x] P to kappa: root finding
- [x] Python: expose `SpinvMF` class
- [x] Python: update pyAMReX particle class (https://github.com/AMReX-Codes/pyamrex/pull/518) OR make spin vector a runtime component in ImpactX (that would be best)
- [x] add benchmark test of spin sampling
- [x] add app docs
- [x] add Python docs
- [x] Restart of spin in `Source` element

Used a [von Mises-Fisher distribution](https://dlwhittenbury.github.io/ds-2-sampling-and-visualising-the-von-mises-fisher-distribution-in-p-dimensions.html) for spin initialization.  This supports an unpolarized beam as a special case.

See also [here](https://sphstat.readthedocs.io/en/latest/sphstat.distributions.html).

Follow-up:
- [ ] Add a simple binary (Bernoulli) distribution as a second option.

cemitch99 committed:
Add spin to particle container.

cemitch99 commented:
The failing tests are expected, because initialization has not been implemented yet.  (That is, the call to `AddNParticles` in `InitDistribution.cpp` needs to be updated, TODO.)

cemitch99 committed:
Add spin sampling algorithm.

cemitch99 commented on src/particles/distribution/SpinvMF.H:
```diff
@@ -0,0 +1,123 @@
+            // Basic transformation of the random variables:
+            amrex::ParticleReal c = std::cos(2_prt*pi*x1);
+            amrex::ParticleReal s = std::sin(2_prt*pi*x1);
+            amrex::ParticleReal t;
```
```suggestion
            amrex::ParticleReal t;
            // The form of the function below is obtained from eq (30) of:
            // D. Frisch and U. D. Hanebeck, "Deterministic Von Mises-Fisher Sampling on the Sphere Using
            // Fibonacci Lattices", IEEE, 2023, DOI: 10.1109/SDF-MFI59545.2023.10361396
```

cemitch99 commented on src/particles/distribution/SpinvMF.H:
```diff
@@ -0,0 +1,123 @@
+         *  This is used for initializing particle spin.  There is a natural
+         *  bijective correspondence between vMF distributions and mean
+         *  (polarization) vectors.
+         *
```
```suggestion
         *
         * The algorithm used here is a simplification of the algorithm described in:
         * C. Pinzon and K. Jung, "Fast Python sampler of the von Mises Fisher distribution", in the special
         * case of the 2-sphere.  Additional references used include:
         * K. V. Mardia and P. E. Jupp, Directional Statistics, Wiley, 1999
         * S. Kang and H-S. Oh, "Novel sampling method for the von Mises-Fisher distribution", Stat.   
         * and Comput. 34, 106 (2024), https://doi.org/10.1007/s11222-024-10419-3
```

cemitch99 replied:
K. V. Mardia and P. E. Jupp, Directional Statistics, Wiley, 1999
S. Kang and H-S. Oh, "Novel sampling method for the von Mises-Fisher distribution", Stat. and Comput.
34, 106 (2024), https://doi.org/10.1007/s11222-024-10419-3

cemitch99 committed:
Add citations for documentation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update citations

ax3l committed:
Spin Distribution Call

cemitch99 committed:
Add conditional form of call to AddNParticles.

cemitch99 commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -242,28 +242,40 @@ namespace impactx
         std::optional<amrex::ParticleReal> bunch_charge,
-        std::optional<amrex::Gpu::DeviceVector<amrex::ParticleReal>> w
+        std::optional<amrex::Gpu::DeviceVector<amrex::ParticleReal>> w,
+        std::optional<amrex::Gpu::DeviceVector<amrex::ParticleReal>> sx,
```
@ax3l Ah - lines 245-246 appear to pose a problem.  How does the compiler know whether the first argument after `bunch_charge` is the (optional) vector `w` or the (optional) vector `sx`?

ax3l replied:
In C++, we do not have named arguments in the location that calls the function, so one would pass either
```C++
AddNParticles(..., bunch_charge, std::nullopt, sx, ...);
```
or
```C++
AddNParticles(..., std::nullopt, w, sx, ...);
```

In the function body of `AddNParticles`, we call `.has_value()` to check if a `std::optional` argument was set.

cemitch99 replied:
Got it; easy fix.

ax3l replied:
Ah, I see you just fixed it in `src/initialization/InitDistribution.cpp` :tada:

cemitch99 commented:
Status and notes:  

1. This builds and compiles locally on a Mac with OpenMP. 
2. The function for sampling the distribution in `SpinvMF` appears to work correctly - this was checked by printing data `(sx,sy,sz)` directly from within `SpinvMF.H`.
3. The use of optional arguments in the call to `AddNParticles` within `InitDistribution.cpp` needs to be clarified.  This causes a problem with initialization of the spin data based on the output from 2.
4. The installation of the Python bindings does not yet work and needs to be fixed.
5. On GPU there is an error triggered by line 406 of `InitDistribution.cpp` (as evidenced by the above tests).

cemitch99 committed:
Add input for a simple distribution test.

ax3l commented:
@cemitch99 thanks, I'll take a look!

ax3l committed:
[Dependencies] pyAMReX: ax3l branch

Update after 25.12 releases to development branch, once
PR is merged
https://github.com/AMReX-Codes/pyamrex/pull/518

cemitch99 committed:
Fix use of optional arguments in AddNParticles call.

ax3l committed:
Formatting

Just some readability, already fixed in prior commit

ax3l committed:
[Dependencies] AMReX: development branch

Update after 25.12 releases to development branch, once
PR is merged
https://github.com/AMReX-Codes/pyamrex/pull/518

cemitch99 committed:
Add P to kappa.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add analysis and README for distribution test.

cemitch99 commented:
Updated status:

A CI test has been added.  This PR builds and compiles locally on a Mac with OpenMP, and the test passes.  Local test:

The following tests FAILED:
	  1 - pytest.AMReX (Failed)
	 34 - chicane.analysis (Failed)
	165 - FODO.tune.py.analysis (Failed)
	175 - chicane_csr.analysis (Failed)
	208 - chicane_isr.analysis (Failed)
	222 - pytest.ImpactX (SEGFAULT)

cemitch99 committed:
Add app/C++ documentation of distribution.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_spin_container

ax3l committed:
Test: Update DataFrame Comparison

ax3l commented:
@WeiqunZhang is it possible that just adding a new ParallelForRNG, even if not using it at runtime, changes the PRNG's sequence or seed? Very puzzled about changes in these unrelated tests.

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -357,6 +367,9 @@ namespace impactx
             amrex::ParticleReal * const AMREX_RESTRICT pt_ptr = pt.data();
+            amrex::ParticleReal * const AMREX_RESTRICT sx_ptr = sx.data();
+            amrex::ParticleReal * const AMREX_RESTRICT sy_ptr = sy.data();
+            amrex::ParticleReal * const AMREX_RESTRICT sz_ptr = sz.data();
```
Could make this a `nullptr` on `!has_spin`, just in case:
https://en.cppreference.com/w/cpp/container/vector/data.html

ax3l commented on examples/distgen/README.rst:
```diff
@@ -219,3 +219,50 @@ We run the following script to analyze correctness:
+.. _examples-distgen-spinvmf:
+
+Spin Sampling from a vMF Distribution
+===============================================
```
```suggestion
Spin Sampling from a vMF Distribution
=====================================
```

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -138,6 +138,16 @@ Initial Beam Distributions
     * ``beam.halo`` (``float``, dimensionless) fraction of charge in halo
 
+Initial Spin Distributions
+---------------------------
```
```suggestion
Initial Spin Distributions
--------------------------
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -551,15 +592,36 @@ namespace impactx
+            // magnitude of the polarization vector (= polarization magnitude)
+            pmag = std::sqrt(polarization_x*polarization_x+polarization_y*polarization_y+polarization_z*polarization_z);
+
+            if (pmag < 1) {
```
See below: I think we need to treat `pmag` == 0 as no spin distribution, do we?

cemitch99 replied:
No, pmag==0 just means the spin distribution is unpolarized (for example, the spin vectors may be distributed uniformly, so the distribution is isotropic).  I would say that there is no spin distribution iff the user does not provide an initial value for any of the spin inputs (e.g., polarization vector components).

ax3l replied:
I see. How do we indicate that we want no spin setting at all now, i.e., what we do for most ImpactX tests?

ax3l replied:
Otherwise the CI was failing for this reason: every test now has a pmag == 0 spin polarization, draws additional random numbers, and thus slightly changes the initial distributions.

Bumping the tolerances would then be the right thing to do.

ax3l replied:
Pushing a commit now for: if the user does not provide at least one polarization component, do not init spin.

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -551,15 +592,36 @@ namespace impactx
             {
                 pp_dist.getWithParser("npart", npart);
-                add_particles(bunch_charge, dist, npart);
+                add_particles(bunch_charge, dist, npart, spin_dist);
```
How do we know there is no `spin_dist` at all? `pmag` is exactly zero?

In that case, we use:
```suggestion
                add_particles(bunch_charge, dist, npart);
```

ax3l commented on src/particles/distribution/SpinvMF.H:
```diff
@@ -0,0 +1,146 @@
+         *
+         * K. V. Mardia and P. E. Jupp, Directional Statistics, Wiley, 1999;
+         * S. Kang and H-S. Oh, "Novel sampling method for the von Mises-Fisher distribution", Stat.
+         * and Comput. 34, 106 (2024), https://doi.org/10.1007/s11222-024-10419-3.
+         *
+         * Return sampling from a vMF distribution.
+         *
+         * @param mux,muy,muz components of a unit vector specifying the mean direction
+         * @param kappa concentration parameter
+         */
+        SpinvMF (
+            amrex::ParticleReal mux,
+            amrex::ParticleReal muy,
+            amrex::ParticleReal muz,
+            amrex::ParticleReal kappa
+        )
+        : m_muX(mux), m_muY(muy), m_muZ(muz), m_kappa(kappa)
+        {
+        }
```
Since we can calculate `kappa` via
```C++
kappa = distribution::SpinvMF::inverse_Langevin(pmag);
```
Do we really need to ask the "SpinvMF user" about it? Or can we do:
```suggestion
        SpinvMF (
            amrex::ParticleReal mux,
            amrex::ParticleReal muy,
            amrex::ParticleReal muz
        )
        : m_muX(mux), m_muY(muy), m_muZ(muz)
        {
            amrex::ParticleReal const pmag = std::sqrt(polarization_x*polarization_x+polarization_y*polarization_y+polarization_z*polarization_z);
            m_kappa = distribution::SpinvMF::inverse_Langevin(pmag);
        }
```

ax3l replied:
Will push this change as well.

ax3l commented on src/particles/distribution/SpinvMF.H:
```diff
@@ -0,0 +1,146 @@
+         */
+        AMREX_GPU_HOST AMREX_FORCE_INLINE
+        static amrex::ParticleReal
+        inverse_Langevin (amrex::ParticleReal const & AMREX_RESTRICT pmag)
```
Just checking I see this correctly (see above):
`inverse_Langevin` handles values between 0 and 1 well, in particular:
- including 0 (result is zero)
- excluding 1 (result would be NaN from div by zero)

?

ax3l replied:
Will add two asserts here.

ax3l committed:
Fix Init (No-Spin Case)

ax3l committed:
rst titles

ax3l commented on src/python/ImpactX.cpp:
```diff
@@ -611,6 +611,7 @@ void init_ImpactX (py::module& m)
         .def("add_particles", &ImpactX::add_particles,
              py::arg("bunch_charge"),
              py::arg("distr"), py::arg("npart"),
+             py::arg("spin_distr") = py::none(),  // TODO: bind the distribution::SpinvMF object to Python
```
Done :) 
```suggestion
             py::arg("spin_distr") = py::none(),
```

ax3l committed:
Remove todo comment

ax3l committed:
Improve `SpinvMF`

ax3l committed:
NVCC: Work-Around Lambda

Same work-around for NVCC as for phase space init.

ax3l committed:
pyAMReX: `development`

ax3l committed:
Python Test

ax3l committed:
`Source` Element: Support Spin in Restart

ax3l commented:
I think the NVCC runner has some issues... Not sure why it aborts.

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_spin_container

ax3l committed:
CUDA CI: Lower Build Parallelism

Aborts close to the end of the build process. Probably
a resource limit hit on the runners.

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -659,6 +660,25 @@ For the input from Twiss parameters in Python, please use the helper function ``
    A 6D stationary thermal or bithermal distribution.
 
+Initial Beam Spin Distribution
+------------------------------
```
Just realized that in this PR, we forgot to update the `add_n_particles` documentation in this file with the new optional parameter signature for spin. Will post a follow-up PR: #1461
